File: stacks_queues/stack_project/lib/stack.js
// ============================================================================
// Implementation Exercise: Stack
// ============================================================================
//
// -------
// Prompt:
// -------
//
// Implement a Stack and all of its methods below!
//
// ------------
// Constraints:
// ------------
//
// Make sure the time and space complexity of each is equivalent to those 
// in the table provided in the Time and Space Complexity Analysis section
// of your Stack reading!
//
// -----------
// Let's Code!
// -----------

class Node {
    constructor(value){
        this.value = value;
        this.next = null;
    }
}

class Stack {
    constructor(){
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    push(val){
        let newNode = new Node(val);

        if(!this.top){
            this.top = newNode;
            this.bottom = newNode;
        } else if(this.length >= 1){
            newNode.next = this.top
            this.top = newNode
        }

        return ++this.length
    }

    pop(){
        if(this.length === 0){
            return null
        } else if(this.length === 1){
            let poppedNode = this.top.value
            this.top = null;
            this.bottom = null;
            this.length--
            return poppedNode;
        } else{
            let poppedNode = this.top.value
            this.top = this.top.next
            this.length--
            return poppedNode;
        }
    }

    size(){
        return this.length
    }
}

exports.Node = Node;
exports.Stack = Stack;

File: miscellenous/data_structure_implementations/stack.py
class Stack:
    def __init__(self):
        self.stack = []
        self.length = 0
    
    def append(self,data):
        self.stack.append(data)
        self.length += 1
        return self


    def pop(self):
        if self.isEmpty(): return

        data = self.stack.pop()
        self.length -= 1
        return data


    def isEmpty(self):
        return self.stack == []


    def size(self):
        return self.length


# stack = Stack()

# stack.append(1)
# stack.append(2)
# stack.append(3)
# stack.append(4)
# print(stack.pop())
# print(stack.pop())
# print(stack.pop())
# print(stack.pop())
# print(stack.size())
# print(stack.pop())
# print(stack.size())
